astropy/utils/data_info.py

```diff
@@ -0,0 +1,303 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# Note: these functions and classes are tested extensively in astropy table
+# tests via their use in providing mixin column info, and in
+# astropy/tests/test_info for providing table and column info summary data.
+
+import os
+import sys
+from copy import deepcopy
+import weakref
+import numpy as np
+from functools import partial
+
+from ..extern import six
+from ..utils import OrderedDict
+from ..utils.compat import NUMPY_LT_1_8
+
+def data_info_factory(names, funcs):
+    """
+    Factory to create a function that can be used as an ``option``
+    for outputting data object summary information.
+
+    Examples
+    --------
+    >>> from astropy.utils.data_info import data_info_factory
+    >>> from astropy.table import Column
+    >>> c = Column([4., 3., 2., 1.])
+    >>> mystats = data_info_factory(names=['min', 'median', 'max'],
+    ...                             funcs=[np.min, np.median, np.max])
+    >>> c.info(option=mystats)
+    min = 1.0
+    median = 2.5
+    max = 4.0
+    n_bad = 0
+    length = 4
+
+    Parameters
+    ----------
+    names: list
+        List of information attribute names
+    funcs: list
+        List of functions that compute the corresponding information attribute
+
+    Returns
+    -------
+    func: function
+        Function that can be used as a data info option
+    """
+    def func(dat):
+        outs = []
+        for name, func in zip(names, funcs):
+            try:
+                if isinstance(func, six.string_types):
+                    out = getattr(dat, func)()
+                else:
+                    out = func(dat)
+            except:
+                outs.append('--')
+            else:
+                outs.append(str(out))
+
+        return OrderedDict(zip(names, outs))
+    return func
+
+
+def _get_data_attribute(dat, attr=None):
+    """
+    Get a data object attribute for the ``attributes`` info summary method
+    """
+    if attr == 'class':
+        val = type(dat).__name__
+    elif attr == 'dtype':
+        val = dat.info.dtype.name
+    elif attr == 'shape':
+        datshape = dat.shape[1:]
+        val = datshape if datshape else ''
+    else:
+        val = getattr(dat.info, attr)
+    if val is None:
+        val = ''
+    return str(val)
+
+
+class InfoDescriptor(object):
```

Comment:
Maybe a simple way to state this pattern is:

> Say what you mean.

So instead of the indirection of:

```python
class MyClass:
    info = InfoDescriptor(DataInfo)
```
which says "info is a thing that somehow when I access it will be a `DataInfo` object".  It forces the reader to have to think about descriptors and how that's all implemented.  Where what you really want is to be able to write:

```python
class MyClass:
    info = DataInfo()
```
which you can read as "info is a `DataInfo`".